#164139684: Enable social login via Google and Facebook

Dubby20 opened the pull request:
#### What does this PR do?
- It uses `passport.js` to authenticate social login

#### Description of Task to be completed?
- Creates a social login authentication so that users will be able to login with their social account.

#### How should this be manually tested?
- Run `npm run start` to start the application.
- Make a GET request to `/api/v1/auth/google` which takes you to a consent screen to allow the application access your social account profile
- Make a GET request to `/api/v1/auth/facebook` which takes you to a consent screen to allow the application access your social account profile

#### Any background context you want to provide?### What are the relevant pivotal tracker stories?
- [#164139684](https://www.pivotaltracker.com/story/show/164139684)

#### Screenshots (if appropriate)
<img width="527" alt="screenshot 2019-03-05 at 12 56 05 pm" src="https://user-images.githubusercontent.com/29398098/54036245-6292fa00-41bb-11e9-80c1-b2c6ca47961d.png">

#### Questions:
- N/A

Commit:
ft(google-login):integrate google authentication
users should be able to login/register with their google account

Commit:
(ft-facebook-login): integrate facebook login
users should be able to login or register with their facebook account

Dubby20 committed:
(ft-twitter-login): integrate twitter login

Dubby20 committed:
(ch-unit-test): write tests that pass

Dubby20 committed:
add wip facebook login

olajide1234 commented on dist/README.md:
```diff
@@ -0,0 +1,394 @@
+Authors Haven - A Social platform for the creative at heart.
```
Hi. I am not sure your dist/ folder should have been uploaded, can you double-check that and your gitignore file. Thanks

kleva-j replied:
I think it is alright to upload your .gitignore file

olajide1234 replied:
Yes, apologies. I meant double check that your .gitignore file currently includes your `dist/` folder among folders to be ignored.

Dubby20 replied:
@olajide1234  and @kleva-j. Thanks for the feedback. The dist folder is ignored in my .gitignore file. However, I didn't edit any of those files but got it into my branch when I pulled from develop because I needed to document my API.

olajide1234 commented on dist/auth/facebook.js:
```diff
@@ -0,0 +1,34 @@
+'use strict';
```
Hi. I am not sure your dist/ folder should have been uploaded, can you double-check that and your gitignore file. Thanks

olajide1234 commented on dist/auth/google.js:
```diff
@@ -0,0 +1,71 @@
+'use strict';
```
Hi. I am not sure your dist/ folder should have been uploaded, can you double-check that and your gitignore file. Thanks

olajide1234 commented on doc.json:
```diff
@@ -45,6 +45,66 @@
+        "produces": [
+          "authorization token"
+        ],
+        "responses": {
```
Please check if you can include parameters required in this Swagger doc.

Dubby20 replied:
Thanks, Jide but I don't have parameters in this feature.

olajide1234 commented on index.js:
```diff
@@ -22,6 +24,19 @@ app.use(morgan('combined'));
 app.use(bodyParser.json());
 
+app.use(session({
+  secret: 'team vidar',
```
Please consider keeping the secret and key in a .env file

Dubby20 replied:
Alright. Thanks

olajide1234 reviewed (commented):
Great job Jacinta. Just look at these few comments. Thanks.

Dubby20 committed:
(ft-social-login): enable social login

users should be able to login/register with their google or facebook account

[Finishes #164139684]

codinger41 commented on auth/facebook.js:
```diff
@@ -0,0 +1,63 @@
+passport.use(new FacebookStrategy({
+  clientID: process.env.FACEBOOK_CLIENT_ID,
+  clientSecret: process.env.FACEBOOK_CLIENT_SECRET,
+  callbackURL: 'http://localhost:7000/api/v1/auth/facebook/callback',
```
@Dubby20  remember to change this callbackURL

Dubby20 replied:
Alright I will but  it has to be on Staging branch

codinger41 commented on .travis.yml:
```diff
@@ -1,7 +1,11 @@
-  - npm install && npm install sequelize-cli -g
-  - npm run test
+ - npm install && npm install sequelize-cli -g
+ - npm run test
```
fix EOL issue

codinger41 commented on auth/google.js:
```diff
@@ -0,0 +1,60 @@
+      where: { email },
+      defaults: {
+        name: profile.displayName,
+        username: profile.name.givenName,
```
Here we can destructure the properties of `profile` so we won't keep repeating `profile`.

codinger41 reviewed (changes requested):
Great job @Dubby20. Please make the few changes I suggested then I can approve.

kleva-j commented on doc.json:
```diff
@@ -45,6 +45,66 @@
         }
       }
+    },
+    "/api/v1/auth/google": {
```
For this google auth route, you should have it as `/auth/google` because a basePath of `/api/v1` has already been set.

Dubby20 replied:
Fixed. Thanks

kleva-j commented on models/User.js:
```diff
@@ -31,7 +31,7 @@ module.exports = (sequelize, DataTypes) => {
     password: {
       type: DataTypes.STRING,
-      allowNull: false,
+      allowNull: true,
```
Just concerned with the reason why you changed `allowNull` to `true`.

kleva-j replied:
I get it now.

Dubby20 committed:
resolve merge conflict

9jaswag commented on auth/passport.js:
```diff
@@ -0,0 +1,109 @@
+import { Strategy as FacebookStrategy } from 'passport-facebook';
+import { User } from '../models';
+
+passport.use(
```
@Dubby20 no fix for this Hound error.

Dubby20 replied:
The problem is that the fix it suggested was throwing eslint error in the code editor

9jaswag commented:
@Dubby20 is this merged in staging? If yes, please add the proper label.

9jaswag commented:
@Dubby20 fix the merge conflicts here.

Dubby20 committed:
fix merge conflict
